src/metrics.rs

```diff
@@ -33,34 +35,33 @@ struct Report {
 }
 
 impl Metrics {
-    pub fn report_error(&mut self, e: (HttpError, UserRequest)) {
-        self.http_errors.lock().unwrap().push(e);
-    }
-
-    pub fn report_request_duration(&mut self, request: (UserRequest, Duration)) {
-        self.request_times.lock().unwrap().push(request);
-    }
-
-    pub fn report_message_sent(&mut self, message_id: String) {
-        self.messages
-            .lock()
-            .unwrap()
-            .entry(message_id)
-            .or_default()
-            .sent = Some(Instant::now());
-    }
-
-    pub fn report_message_received(&mut self, message_id: String) {
-        self.messages
-            .lock()
-            .unwrap()
-            .entry(message_id)
-            .or_default()
-            .received = Some(Instant::now());
+    pub fn new() -> (Self, Sender<Event>) {
+        let http_errors: Arc<Mutex<Vec<(UserRequest, HttpError)>>> = Arc::new(Mutex::new(vec![]));
+        let request_times: Arc<Mutex<Vec<(UserRequest, Duration)>>> = Arc::new(Mutex::new(vec![]));
+        let messages: Arc<Mutex<HashMap<String, MessageTimes>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
+        let (tx, rx) = mpsc::channel::<Event>(100);
+        let receiver = Arc::new(Mutex::new(rx));
+
+        tokio::spawn(read_events(
```

Comment:
nit: Not really important right now, but could we have the spawn be part of a different, say `run`, method in Metrics to be called by the client instead of doing it implicitly in the constructor?